oak_attestation_verification/src/amd.rs

```diff
@@ -0,0 +1,202 @@
+const CHIP_ID_OID: ObjectIdentifier = ObjectIdentifier::new_unwrap("1.3.6.1.4.1.3704.1.4");
+const UCODE_OID: ObjectIdentifier = ObjectIdentifier::new_unwrap("1.3.6.1.4.1.3704.1.3.8");
+
+// Verifies validity of a matching ARK, ASK certificate pair.
+//
+// Validate at least a subset of Appendix B.3 of
+// https://www.amd.com/content/dam/amd/en/documents/epyc-technical-docs/programmer-references/55766_SEV-KM_API_Specification.pdf
+// Ideally, we'd check everything listed there.
+pub fn validate_ark_ask_certs(ark: &Certificate, ask: &Certificate) -> anyhow::Result<()> {
+    if ark.tbs_certificate.version != Version::V3 {
+        anyhow::bail!("unexpected version of ARK cert");
+    }
+
+    if ask.tbs_certificate.version != Version::V3 {
+        anyhow::bail!("unexpected version of ASK cert");
+    }
+
+    verify_cert_signature(ark, ask)?;
+    verify_cert_signature(ark, ark)?;
+
+    Ok(())
+}
+
+pub fn verify_cert_signature(signer: &Certificate, signee: &Certificate) -> anyhow::Result<()> {
+    if signee.signature_algorithm.oid != RSA_SSA_PSS_OID {
+        anyhow::bail!(
+            "unsupported signature algorithm: {:?}",
+            signee.signature_algorithm
+        );
+    }
+
+    let verifying_key = {
+        let pubkey_info = signer
+            .tbs_certificate
+            .subject_public_key_info
+            .owned_to_ref();
+        let pubkey = RsaPublicKey::try_from(pubkey_info)
+            .map_err(|_err| anyhow::anyhow!("could not parse RSA public key"))?;
+        rsa::pss::VerifyingKey::<Sha384>::new(pubkey)
+    };
+
+    let message = signee
+        .tbs_certificate
+        .to_der()
+        .map_err(|_err| anyhow::anyhow!("could not extract message to verify RSA signature"))?;
+    let signature = Signature::try_from(signee.signature.raw_bytes())
+        .map_err(|_err| anyhow::anyhow!("could not extract RSA signature"))?;
+
+    Ok(verifying_key
+        .verify(&message, &signature)
+        .map_err(|_err| anyhow::anyhow!("signature verification failed"))?)
+}
+
+fn to_array_64<T>(slice: &[T]) -> anyhow::Result<&[T; 64]> {
+    if slice.len() == 64 {
+        let ptr = slice.as_ptr() as *const [T; 64];
+        unsafe { Ok(&*ptr) }
+    } else {
+        anyhow::bail!("unexpected length of chip ID")
+    }
+}
+
+// Currently unused.
+pub fn product_name(vcek: &Certificate) -> anyhow::Result<String> {
+    for ext in vcek
+        .tbs_certificate
+        .extensions
+        .as_ref()
+        .ok_or_else(|| anyhow::anyhow!("could not get extensions from VCEK cert"))?
+    {
+        if ext.extn_id == PRODUCT_NAME_OID {
+            return String::from_utf8(ext.extn_value.as_bytes().to_vec())
+                .map_err(|_utf8_err| anyhow::anyhow!("failed to read product name"));
+        }
+    }
+    Err(anyhow::anyhow!("no product name found in VCEK vert"))
+}
+
+fn chip_id(vcek: &Certificate) -> anyhow::Result<[u8; 64]> {
```

Comment:
I think this can just return a byte slice. If the compiler doesn't like the lifetimes it could return Vec. It should be possible to compare the vector to the array using slices over both.